bench/Main.hs

```diff
@@ -0,0 +1,285 @@
+
+exampleModulePath :: FilePath
+exampleModulePath = "Distribution" </> "Simple.hs"
+
+data Config = Config
+  { verbose :: !Bool,
+    -- For some reason, the Shake profile files are truncated and won't load
+    shakeProfiling :: !(Maybe FilePath),
+    outputCSV :: !Bool
+  }
+
+type HasConfig = (?config :: Config)
+
+configP :: Parser Config
+configP = Config
+    <$> (not <$> switch (long "quiet"))
+    <*> optional (strOption (long "shake-profiling" <> metavar "PATH"))
+    <*> switch (long "csv")
+
+output :: (MonadIO m, HasConfig) => String -> m ()
+output = if verbose ?config then liftIO . putStrLn else (\_ -> pure ())
+
+---------------------------------------------------------------------------------------
+
+type Experiment = TextDocumentIdentifier -> Session Bool
+
+data Bench = forall setup.
+  Bench
+  { name :: !String,
+    samples :: !Natural,
+    benchSetup :: TextDocumentIdentifier -> Session setup,
+    experiment :: setup -> Experiment
+  }
+
+bench :: String -> Natural -> Experiment -> Bench
+bench name samples userExperiment = Bench {..}
+  where
+    experiment () = userExperiment
+    benchSetup _ = return ()
+
+benchWithSetup ::
+  String ->
+  Natural ->
+  (TextDocumentIdentifier -> Session p) ->
+  (p -> Experiment) ->
+  Bench
+benchWithSetup = Bench
+
+runBenchmarks :: HasConfig => [Bench] -> IO ()
+runBenchmarks benchmarks = do
+  results <- forM benchmarks $ \b -> (b,) <$> runBench b
+
+  forM_ results $ \(Bench {name, samples}, duration) ->
+    output $
+      "TOTAL "
+        <> name
+        <> " = "
+        <> showDuration duration
+        <> " ("
+        <> show samples
+        <> " repetitions)"
+
+  when (outputCSV ?config) $ do
+    putStrLn $ intercalate ", " $ map name benchmarks
+    putStrLn $ intercalate ", " $ map (showDuration . snd) results
+
+runBench :: HasConfig => Bench -> IO Seconds
+runBench Bench {..} = handleAny (\e -> print e >> return (-1))
+  $ runSessionWithConfig conf cmd lspTestCaps dir
+  $ do
+    doc <- openDoc exampleModulePath "haskell"
+    void (skipManyTill anyMessage message :: Session WorkDoneProgressEndNotification)
+
+    liftIO $ output $ "Running " <> name <> " benchmark"
+    userState <- benchSetup doc
+    let loop 0 = return True
+        loop n = do
+          (t, res) <- duration $ experiment userState doc
+          if not res
+            then return False
+            else do
+              output (showDuration t)
+              loop (n -1)
+
+    (t, res) <- duration $ loop samples
+
+    exitServer
+    -- sleeep to give ghcide a chance to print the RTS stats
+    liftIO $ threadDelay 50000
+
+    return $ if res then t else -1
+  where
+    cmd =
+      unwords $
+        [ "ghcide",
+          "--lsp",
+          "--cwd",
+          dir,
+          "+RTS",
+          "-N4",
+          "-S",
+          "-RTS"
+        ]
+          ++ concat
+            [ ["--shake-profiling", path]
```

Comment:
`--shake-profiling` can add a significant overhead to benchmarking. Use with caution I think.

Reply:
Agreed. FWIW, the majority of time in profiling goes to writing the output file, which is sadly an unavoidable step.

Reply:
It's turned off by default